add sats_per_kweight option when crafting a transaction

ziggie1984 opened the pull request:
Replaces #7443 because a wrong `from` branch was used.

For the commands `SendCoins`, `SendMany`, `CloseChannel`, `OpenChannel`, `BatchOpenChannel` and in the response of the `EstimateFee` command the option `sats_per_kweight` is added to allow for more granular control of fees when creating transactions. 

In the current fee environment one cannot use decimal fee estimates for sats_per_vbyte, therefore using sats_per_kweight one will be able to use decimal sats_per_vbyte feerates indirectly without introducing floating point math.

Currently in the lnd backend the option sats_per_kweight is prioritized in case both options are present at the same time (sats_per_kweight and sats_per_vbyte). Though having both of them set is prevented on `lncli` level. This might be open for discussion. 

## Steps to Test
Changes where done on the rpc level and `lncli` therefore Polar was used to test all the changed commands.

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

joostjager commented on lnrpc/lightning.proto:
```diff
@@ -2000,6 +2015,14 @@ message CloseChannelRequest {
+
+    NOTE: This field is only respected if we're the initiator of the channel.
+    */
+    uint64 sat_per_kweight = 8;
```
Deprecate old field?

ziggie1984 replied:
you mean instead of the `sat_per_byte` use `sat_per_kweight` ?

joostjager replied:
yes

joostjager commented on rpcserver.go:
```diff
@@ -245,6 +246,11 @@ func calculateFeeRate(satPerByte, satPerVByte uint64, targetConf uint32,
 	}
 
+	// Default to satPerVByte, and overwrite it if satPerKWeight is set.
+	if satPerKWeight != 0 {
```
I'd make this more strict and return an error if both are set.

alexbosworth commented:
`FundPsbt` also uses `sats_per_vbyte`

ziggie1984 commented:
takes a bit longer because I want to add itests for all the functions to easily test that changes work.

ziggie1984 commented:
I need some feedback, I wanna test for example `openchannel` with a specific feerate but I am not sure if my approach is fine because I am already anticipating too much regarding the EDDSA overEstimate? 

```
// testOpenChannelFeerate tests whether a channel opens with the specified
// feerate.
//
// NOTE: Feerates cannot be compared because they are rounded down and the
// fee estimation accounts for some additonal weight units which need to be
// taken into account. Therefore fees are compared rather than feerate.
func testOpenChannelFeerate(ht *lntest.HarnessTest) {

	alice, bob := ht.Alice, ht.Bob
	chanAmt := funding.MaxBtcFundingAmount
	feerateSatPerKweight := btcutil.Amount(1000)
	openChannelReq := lntest.OpenChannelParams{
		Amt:           chanAmt,
		SatPerKWeight: feerateSatPerKwe<ight,
	}

	// Create a channel Alice->Bob.
	chanPoint := ht.OpenChannel(alice, bob, openChannelReq)
	defer ht.CloseChannel(alice, chanPoint)

	fundingTxID := ht.GetChanPointFundingTxid(chanPoint)
	rawTx := ht.Miner.GetRawTransaction(fundingTxID).MsgTx()
	txFee := int64(ht.CalculateTxFee(rawTx))
	txWeight := ht.CalculateTxWeight(rawTx)

	weightCorrection := txWeight + ht.CalculateECDSASigOverEstimate(rawTx)

	// Round down as required by Bolt03
	expectedFee := weightCorrection * int64(feerateSatPerKweight) / 1000

	require.Equal(ht, txFee, expectedFee)

}
```

ziggie1984 commented:
Still some open questions I have:

1. Currently I replaced the deprecated option `sat_per_byte` and added the new option `sat_per_kweight` in one commit, is this fine or should this be done in different? 

2. I did not include the itest for `bumpfee` for now, I guess it also makes sense to to this ? 

3. I added the `sat_per_kweight` option for almost all commands/function except for the watchertower feerate. I did not find it that important to include it there as well, any thoughts ?

4. Regarding the tests I choose to compare the result including an error-rate (`inepsilon`) as seen in other place in the code. I choose 1% error tolerance. Can we go with this approach ?

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

ziggie1984 commented:
This PR needs some rework from my side because I did renumber protofile nummeration not thinking of backwards compatibility issues. So will rework in the next days.

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@ziggie1984, remember to re-request review from reviewers when ready

ziggie1984 commented:
I updated this PR to make the RPC chane backwards-compatible. I included itests for commands which now have the new sat_per_kweight option. 

Only for `estimatefee` cmd where I added the sat_per_kweight option I did not include a test, because nothing really changed compared to the sat_per_vbyte option.

From my POV this PR is ready for review. Don't be intimidated by the amount of lines which were changed. The change is minor but the walletrpc and lnrpc commands changed therefore the change appears to be big.

lightninglabs-deploy commented:
@joostjager: review reminder

lightninglabs-deploy commented:
@joostjager: review reminder

lightninglabs-deploy commented:
@joostjager: review reminder

lightninglabs-deploy commented:
@joostjager: review reminder

lightninglabs-deploy commented:
@joostjager: review reminder

lightninglabs-deploy commented:
@joostjager: review reminder

lightninglabs-deploy commented:
@joostjager: review reminder

lightninglabs-deploy commented:
@joostjager: review reminder

ziggie1984 commented:
!lightninglabs-deploy mute

C-Otto commented:
FYI: I'd love to see this merged. Currently, it has a bunch of merge conflicts.

joostjager reviewed (commented):
PR has been untouched for a long time, but cannot remove myself as reviewer.

starius commented on lnrpc/lightning.proto:
```diff
@@ -1111,9 +1111,8 @@ message EstimateFeeResponse {
-    // The fee rate in satoshi/vbyte.
-    int64 feerate_sat_per_byte = 2 [deprecated = true];
+    // The fee rate in satoshi/kweight.
+    int64 sat_per_kweight = 2;
```
I propose to use another ID for this field and not to change the deprecated field, not to break binary backwards compatibility. If someone has an old client and runs against new server (or vice-versa), they can send a request with this field, which has changed meaning, resulting in huge overpayment or underpayment of fee.
